lib/Target/LLVM/LLVMToDynamicLib.cpp

```diff
@@ -0,0 +1,303 @@
+
+#include "ttmlir/Target/LLVM/LLVMToDynamicLib.h"
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Target/LLVMIR/ModuleTranslation.h"
+
+// #include "lld/Common/Driver.h"
+// #include "lld/Common/ErrorHandler.h"
+// #include "lld/Common/Memory.h"
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/ExecutionEngine/ExecutionEngine.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Target/TargetOptions.h"
+#include "llvm/TargetParser/Host.h"
+
+// a lot of this code is adapted from IREE's LLVMCPU target code
+namespace mlir::tt::llvm_to_cpu {
+
+// Function to convert MLIR ModuleOp to LLVM Module
+std::unique_ptr<llvm::Module>
+convertToLLVMModule(mlir::ModuleOp mlirModule, llvm::LLVMContext &llvmContext) {
+  // Ensure the MLIR module is in the LLVM dialect
+  if (!mlirModule.getOperation()
+           ->hasTrait<mlir::OpTrait::IsIsolatedFromAbove>()) {
+    llvm::errs() << "ModuleOp is not properly isolated\n";
+    return nullptr;
+  }
+
+  // Use MLIR's translation utility
+  auto llvmModule = mlir::translateModuleToLLVMIR(mlirModule, llvmContext,
+                                                  "test-llvm-custom-name");
+  if (!llvmModule) {
+    llvm::errs() << "Failed to convert MLIR ModuleOp to LLVM IR\n";
+    return nullptr;
+  }
+
+  return llvmModule;
+}
+
+std::unique_ptr<llvm::TargetMachine>
+createTargetMachine(const std::string &targetTriple) {
+  std::string errorMessage;
+  auto llvmTarget =
+      llvm::TargetRegistry::lookupTarget(targetTriple, errorMessage);
+  if (!llvmTarget) {
+    llvm::errs() << "target lookup failed for " << targetTriple
+                 << " w msg: " << errorMessage << "\n";
+    return nullptr;
+  }
+
+  llvm::TargetOptions options;
+
+  std::unique_ptr<llvm::TargetMachine> machine(llvmTarget->createTargetMachine(
+      targetTriple, "generic" /* cpu e.g k8 */,
+      "" /* cpu features e.g avx512f */, options, llvm::Reloc::Model::PIC_));
+  return machine;
+}
+
+llvm::LogicalResult compileToObject(llvm::Module &module,
+                                    llvm::LLVMContext &context,
+                                    const std::string &outputFilename) {
+  // Initialize LLVM targets
+  LLVMInitializeX86Target();
+  LLVMInitializeX86TargetMC();
+  LLVMInitializeX86TargetInfo();
+  LLVMInitializeX86AsmPrinter();
+  LLVMInitializeX86AsmParser();
+
+  llvm::errs()
+      << "(debug) Registered targets after explicit x86 registration :\n";
```

Comment:
Let's remove the stray prints before merging.

Reply:
Yep, sorry